### PR TITLE
BUG: Use new col names and fill_values together while read ascii table

### PR DESCRIPTION
**Pull request by MridulS**
This is an attempt to fix https://github.com/astropy/astropy/issues/13851 but this will fail tests added in https://github.com/astropy/astropy/pull/11593. I am not sure if I understand the tests added in https://github.com/astropy/astropy/pull/11593, it adds a new table for the testcase with duplicate labels (`EW`). This creates an ambiguous case with the internal `fill_values` in https://github.com/astropy/astropy/blob/ae438b4ad4cb14355d2efc4acc0477b22068d83f/astropy/io/ascii/cds.py#L190 which creates `('-9.9', 'nan', 'EW'), ('', 'nan', 'EW')` and these are supposed to be different columns (right?).

Before this change the test was passing as it just created an extra `EW_1` column and ran both of those fill_value filters on `EW` before doing that.

I feel like I am probably missing something here 😅

### CODE REVIEW

**Comment by hamogu** [comment]
I think this fix looks good and is as it should be. So, why are the tests failing? It's probably a bug that they passed before...  but I'm not sure. I'm looking into it more. Fortunately, just an hour ago, I started writing a new table reader so my head is in this code right now. Good timing!

File: astropy/io/ascii/tests/test_fixedwidth.py
Comment (hamogu) [inline astropy/io/ascii/tests/test_fixedwidth.py:262]:
```suggestion
    assert dat[2][2] is np.ma.masked
```

**Comment by taldcroft** [comment]
@MridulS - I took a first look at this and your fix seems quite nice and simple. But we do need to be careful about this change given the sometimes-twisted interactions in io.ascii (like the CDS case).

About the CDS test, that file was modified incorrectly as you note to include multiple `EW` columns. I think it is fair to fix that CDS test here. I do not believe that a real CDS/MRT file can have multiple columns with the same name. But maybe @hamogu has thoughts.

**Comment by hamogu** [comment]
@taldcroft I consider this a bug (there is no reason why the internal column names that get discarded before the user even sees them should be used to label the fill_values), but I wonder if we should treat is as as API change and milestone it for 7.0? It might potentially break user code.

@MridulS I agree with Tom just fix the test. All CDS tables that I have seen have unique column names. I don't know if that's formally a requirement, but in practice it is.

**Comment by pllim** [comment]
OK I removed the backport label. If you do want it in 7.0 instead of 6.1, please update the milestone again and turn this into draft because current `main` is slated for v6.1, not v7.0 yet. Thanks, all!

**Comment by taldcroft** [comment]
> but I wonder if we should treat is as as API change and milestone it for 7.0? It might potentially break user code.

I agree this can potentially break some code, but only code that was using undocumented behavior which we agree is a bug. At some level the original behavior was not part of the API. So I would recommend 6.1 for release with an `15774.api.rst` file which explains this corner case.

APE-21 says *Minimize as much as possible any changes that break API or change results in minor releases, but be pragmatic and occasionally allow such changes if they would not affect many users*. I think this PR fits that description and it would be shame to have this bug fix sit around for 1.5 years.

**Comment by hamogu** [comment]
I guess I'm falling into backspace heating: https://xkcd.com/1172/

**Comment by dhomeier** [comment]
I agree that test should be fixed, though perhaps not to the effect of invalidating columns with duplicate labels (which would require some additional check in the reader code anyway). Those columns may seem awkward, but looking at the [Readme specifications](http://vizier.u-strasbg.fr/doc/catstd-3.1.htx) I don't see any clear rule that the `Label` entries have to be unique. Many of the example values are more descriptive of a physical type or similar, like `RAh`, `plx`, `Vmag`, that could well occur multiple times in a table (e.g. for binary stars, Zeeman components etc.) I think the whole format does not really involve the concept of a `Table` object with uniquely named columns.
What the test table + Readme1 are read in as is
```python
<Table length=2>
Cluster  Star   Wave   El   ion   chiEx   loggf     EW     EW_1    e_EW     Q   
               0.1 nm               eV            0.1 pm  0.1 pm  0.1 pm        
  str5  int64 float64 str2 int64 float64 float64 float64 float64 float64 float64
------- ----- ------- ---- ----- ------- ------- ------- ------- ------- -------
  Cr110  2108 6696.79   Al     1    4.02   -1.42    29.5      --     2.2   0.289
  Cr110  2108 6698.67   Al     1    3.14   -1.65    58.0      --     2.0   0.325
```
i.e. we obviously already have functionality to make column names unique where they aren't yet, and that functionality probably should not be broken (and in fact might make sense to be tested explicitly). The problem with the table is rather that the extra column contains invalid entries only, as the manipulated header is pointing it to empty columns in the ascii file:
```
  39- 43  F5.1  0.1pm     EW        ?=-9.9 Equivalent width (in mA)
  44- 45  F5.1  0.1pm     EW        [1/52]?-= Equivalent width (in mA)
  46- 49  F4.1  0.1pm   e_EW        [1/6]?+ Luminosity class codified (11)
```
Note that there is also a mismatch between that column's `Bytes` length and format `F5.1`, which is perhaps something that should rather be validated. Though maybe we can (continue to) rely on anything having made it into CDS to have properly validated headers...

Can't tell yet exactly _why_ the test was working before, but I suppose it was using some default `fill_values` for the columns applying to its generic name `col7` or whatever that accepted the blank input; apparently the CDS Readme names are applied after the fact similar to the custom `names` specified to `read()`. So with this PR  those `fill_values` would probably have to be specified specifically for the `EW_1` column. I have not succeeded to find a setting that works so far (actually could not find any documentation for the exact format/syntax of the `fill_values` parameter).
Alternatively, and maybe better, we could fix the table body to include valid numbers for the additional `EW` column (and fix its `Bytes` range). `test_cds_no_whitespace` has no checks for actual column data afaics, so I have no idea if there ever was an intention to test empty/missing data values here.

**Comment by MridulS** [comment]
> The problem with the table is rather that the extra column contains invalid entries only, as the manipulated header is pointing it to empty columns in the ascii file

This in indeed the issue, but there is currently no validation check of the header file with the data file. Are there tools already out there which do this validation?

In the latest commit I have removed the duplicated column from the Readme file description and edited the tests.

I would argue this is not an API breakage as this test itself depends on invalid data? There is no data for the duplicated column `EW` in the data file `table.dat`.

**Comment by taldcroft** [comment]
As an aside, there is something that is confusing me. The way that the CDS reader is called seems to matter. This is using astropy with the patch here in #15774 applied but with the original `ReadMe1` file in `main`.
```
from astropy.io import ascii
from pathlib import Path

readme = "astropy/io/ascii/tests/data/cds/null/Readme1"
data = "astropy/io/ascii/tests/data/cds/null/table.dat"

# THIS WORKS
t = ascii.read(data, readme=readme, format="cds")
print(t)
# Cluster Star   Wave   El ion chiEx loggf   EW    EW_1   e_EW    Q
#               0.1 nm           eV        0.1 pm 0.1 pm 0.1 pm
# ------- ---- ------- --- --- ----- ----- ------ ------ ------ -----
#   Cr110 2108 6696.79  Al   1  4.02 -1.42   29.5     --    2.2 0.289
#   Cr110 2108 6698.67  Al   1  3.14 -1.65   58.0     --    2.0 0.325

# THIS DOES NOT WORK
r = ascii.Cds(readme)
r.read(data)
# ValueError: Column EW_1 failed to convert: could not convert string to float: ''
```

**Comment by MridulS** [comment]
Interesting, this means they are taking different code paths (???), need to investigate this further.

On this tangent, is this the recommended way of reading files:
``` python
t = ascii.read(data, readme=readme, format="cds")
```
instead of?
``` python
r = ascii.Cds(readme)
r.read(data)
```

**Comment by taldcroft** [comment]
Overall, the CDS reader is fragile and already broken in the case of multiple columns with the same name. That was noted by @MridulS in the initial description and I confirm that. The `fill_values` that the CDS reader generates use the non-unique name, so no matter what things are just broken.

My takeaway is that this patch is fine, where we just change that one `EW` to `EW_1` in `ReadMe1` to make the columns unique. Having unique column names conforms to the reality of CDS in practice, even if their "specification" is not specific about the point [^1]. So basically we should not spend any effort/time worrying about a case that does not happen in the real world.

[^1]: If you can even call providing one example as a "specification".

**Comment by taldcroft** [comment]
Yes, `t = ascii.read(data, readme=readme, format="cds")` is the recommended "modern" way.

**Comment by taldcroft** [comment]
@MridulS - FYI `io.ascii` is mostly very old code, much of written when I had only known Python for a year or two before astropy even existed (see https://pypi.org/project/asciitable/). It's really over-complicated and I would do things very differently today. Sigh.

**Comment by dhomeier** [comment]
> Yes, `t = ascii.read(data, readme=readme, format="cds")` is the recommended "modern" way. Or even more commonly, `t = Table.read(data, readme=readme, format="ascii.cds")`.

So should the old uses of `get_reader`, `ascii.Cds` etc. be replaced at some point, or at least annotated that they are still using a deprecated interface? I've been wondering myself why the tests were using one form or the other, and struggled finding a way to get `fill_values` added to `ascii.Cds`.

File: astropy/io/ascii/tests/data/cds/null/ReadMe1
Comment (dhomeier) [inline astropy/io/ascii/tests/data/cds/null/ReadMe1:65]:
```suggestion
  44- 48  F5.1  0.1pm     EW_1      [1/52]?-= Another equivalent width (in mA)
  49- 52  F4.1  0.1pm   e_EW        [1/6]?+ Luminosity class codified (11)
  53- 59  F6.3  ---       Q         ?=-9.999 DAOSPEC quality parameter Q
```
I found now that the assigning of unique column names is automatically done in https://github.com/astropy/astropy/blob/651ea27844801b19895b87e6ec7f776b1fd39a61/astropy/io/ascii/core.py#L1292-L1294
so this is not specific to CDS and there is probably no need to test the duplicate label here.
But I'd still vote for fixing the format specification (this would require a new `table1.dat` that actually has the extra columns and then might actually include some data so the _invalid input for `float`_ would not occur in the first place...)
Alternatively could perhaps change the format to `F2.0`, but that does not sound very sensible either.

Comment (MridulS) [reply astropy/io/ascii/tests/data/cds/null/ReadMe1]: These changes still break the tests.

```
test_cds_no_whitespace - ValueError: Column e_EW failed to convert: could not convert string to float: '2  0'
test_cds_order - ValueError: Column e_EW failed to convert: could not convert string to float: '2  0'
```

Comment (dhomeier) [reply astropy/io/ascii/tests/data/cds/null/ReadMe1]: See comment above, you will need to create a new `table1.dat` with a matching data section, e.g. insert something in positions 46-48 like
```
Cr110   2108 6696.79 Al1  4.02  -1.42  29.5 -8.5 2.2  0.289
```
(I'd suggest to use valid numbers, although if you still wish to test the `fill_values` one row could be left empty).

File: astropy/io/ascii/tests/data/cds/null/table1.dat
Comment (dhomeier) [inline astropy/io/ascii/tests/data/cds/null/table1.dat:2]:
```suggestion
Cr110   2108 6698.67 Al1  3.14  -1.65  58.0  -1.3  2.0  0.325
Cr110   2108 6696.79 Al1  4.02  -1.42  29.5  -8.5  2.2  0.289
```
Seems you have missed one column there, though for the tests it works as the actual numbers don't use the full format width.

**Comment by taldcroft** [comment]
> So should the old uses of get_reader, ascii.Cds etc. be replaced at some point

This is a little tricky since this interface is still in the public API, so we should continue testing them. I suppose we could make `get_reader()` be a thin wrapper around `_get_reader()`. Then use `_get_reader()` internally and deprecate `get_reader()` in the public API. 

It will end up being a fair bit of work and churn for no real change in the end except a somewhat cleaner API.

**Comment by MridulS** [comment]
I believe this PR is ready to go?

We can create a new issue to track the future of a cleaner API interface.

**Comment by hamogu** [comment]
@dhomeier and @taldcroft had requests for changes, so I'll let them decide. For now, this is still missing a changelog entry: https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#add-a-changelog-entry

**Comment by dhomeier** [comment]
My suggestions have all been addressed; moving the fate of `get_reader` to a separate issue is a sensible solution.
The commits should be squashed, with the changelog entry this would be ready on my part.

**Review by hamogu — APPROVED** [review]
It seems that somehow we dropped the ball on this when the discussion got carried away for a new cleaner interface etc,
I think we agree that the actual change in the this PR fixes an actual bug and should go in. However, it's been a few months now, so I'm just approving now and give the  other reviewers time to have another look before we press merge.

**Comment by hamogu** [comment]
Mmm. The required tests seem to be stuck.

**Comment by pllim** [comment]
I think need rebase?

**Comment by hamogu** [comment]
I'm trying that. At the very least, it'll re-run  the tests.

**Pull request merged**
